### ISSUE TITLE
pipelining doesn't work with python3.6 on Mac

### ISSUE THREAD

**Issue by mildebrandt** [ansible/ansible#34727]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

sudo

##### ANSIBLE VERSION

```
ansible 2.4.2.0
  config file = ./ansible.cfg
  configured module search path = ['./library', './third-party/kubespray/library']
  ansible python module location = /usr/local/lib/python3.6/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.1 (default, Jun 16 2017, 16:42:41) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]
```

##### CONFIGURATION

```
ANSIBLE_PIPELINING(./ansible.cfg) = True
ANSIBLE_SSH_ARGS(./ansible.cfg) = -o ControlMaster=auto -o ControlPersist=30m -o ConnectionAttempts=100 -o UserKnownHostsFile=/dev/null
DEFAULT_CALLBACK_WHITELIST(./ansible.cfg) = ['profile_tasks']
DEFAULT_GATHERING(./ansible.cfg) = smart
DEFAULT_MODULE_PATH(./ansible.cfg) = ['./library', './third-party/kubespray/library']
DEFAULT_ROLES_PATH(./ansible.cfg) = ['./roles', './third-party/kubespray/roles', '/Users/chr
DEFAULT_STDOUT_CALLBACK(./ansible.cfg) = skippy
DEPRECATION_WARNINGS(./ansible.cfg) = False
HOST_KEY_CHECKING(./ansible.cfg) = False
```

##### OS / ENVIRONMENT

```
ProductName:    Mac OS X
ProductVersion: 10.13.2
BuildVersion:   17C88
```

##### SUMMARY

When running sudo through ssh with pipelining enabled, the selector doesn't detect any events. Inside ansible/plugins/connection/ssh.py I can see that it correctly detected the become_prompt and sent the password to the stdin. Then it sits at `events = selector.select(timeout)` until the timeout. The next line is never sent/never detected from stdout. 

This works with pipelining disabled on python3.6. This also works on python 2.7 with or without pipelining. 

Additionally, the error message `waiting for privilege escalation prompt` isn't correct in this instance. We saw the prompt at this time and it's now waiting for the privilege escalation success message (BECOME-SUCCESS-*).

##### STEPS TO REPRODUCE

I'm just gathering the facts for this test:
```
ANSIBLE_BECOME_PASSWORD=not_my_password ansible -m setup k8s1 -i ~/workspace/kubespray/inventory.cfg -b -v -u k8s --private-key=~/.ssh/id_rsa
```

With the following inventory:

```
k8s1     ansible_host=192.168.56.101 ip=192.168.56.101 ansible_become_pass="{{ lookup('env', 'ANSIBLE_BECOME_PASSWORD') }}"
```

##### EXPECTED RESULTS

I expected the sudo command to succeed.

##### ACTUAL RESULTS

Timeout occurred. 

```
$ ANSIBLE_BECOME_PASSWORD=not_my_password ansible -m setup k8s1 -i a.cfg -b -v -u k8s --private-key=~/.ssh/id_rsa -vvvvv
ansible 2.4.2.0
  config file = ./ansible.cfg
  configured module search path = ['./library', './third-party/kubespray/library']
  ansible python module location = /usr/local/lib/python3.6/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.1 (default, Jun 16 2017, 16:42:41) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]
Using ./ansible.cfg as config file
setting up inventory plugins
Parsed ./a.cfg inventory source with ini plugin
Loading callback plugin minimal of type stdout, v2.0 from /usr/local/lib/python3.6/site-packages/ansible/plugins/callback/__init__.py
META: ran handlers
<192.168.56.101> ESTABLISH SSH CONNECTION FOR USER: k8s
<192.168.56.101> SSH: ansible.cfg set ssh_args: (-o)(ControlMaster=auto)(-o)(ControlPersist=30m)(-o)(ConnectionAttempts=100)(-o)(UserKnownHostsFile=/dev/null)
<192.168.56.101> SSH: ANSIBLE_HOST_KEY_CHECKING/host_key_checking disabled: (-o)(StrictHostKeyChecking=no)
<192.168.56.101> SSH: ANSIBLE_PRIVATE_KEY_FILE/private_key_file/ansible_ssh_private_key_file set: (-o)(IdentityFile="~/.ssh/id_rsa")
<192.168.56.101> SSH: ansible_password/ansible_ssh_pass not set: (-o)(KbdInteractiveAuthentication=no)(-o)(PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey)(-o)(PasswordAuthentication=no)
<192.168.56.101> SSH: ANSIBLE_REMOTE_USER/remote_user/ansible_user/user/-u set: (-o)(User=k8s)
<192.168.56.101> SSH: ANSIBLE_TIMEOUT/timeout set: (-o)(ConnectTimeout=10)
<192.168.56.101> SSH: found only ControlPersist; added ControlPath: (-o)(ControlPath=~/.ansible/cp/38004cf33b)
<192.168.56.101> SSH: EXEC ssh -vvv -o ControlMaster=auto -o ControlPersist=30m -o ConnectionAttempts=100 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o 'IdentityFile="~/.ssh/id_rsa"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=k8s -o ConnectTimeout=10 -o ControlPath=~/.ansible/cp/38004cf33b 192.168.56.101 '/bin/sh -c '"'"'sudo -H -S  -p "[sudo via ansible, key=ocoqtxdwvgdhofoilcidzwuxjmwrozum] password: " -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-ocoqtxdwvgdhofoilcidzwuxjmwrozum; /usr/bin/python'"'"'"'"'"'"'"'"' && sleep 0'"'"''
k8s1 | FAILED! => {
    "msg": "Timeout (12s) waiting for privilege escalation prompt: "
}
```

**Comment by ansibot** [comment]
Files identified in the description:
None

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by foozmeat** [comment]
I've been experiencing this issue as well since moving to python3 on my mac. Toggling pipelining gets it working again.

**Comment by mildebrandt** [comment]
@foozmeat Can you describe what you mean by "toggling"? I don't understand it in this context. That usually means setting it to False and then setting it back to True. I don't understand how that would fix this issue. Thanks.

**Comment by foozmeat** [comment]
@eyeofthefrog yes you're right, I made a confusing comment. In the past we've had pipelining set to True but have to set it to False if ansible is running under python3 or else we get the error described above.

**Comment by mildebrandt** [comment]
@foozmeat I see, you meant "toggle to off". I should have understood that, sorry. Thanks for clarifying.

**Comment by sivel** [comment]
I wonder if we should instead, just utilize `selectors2` for py2 and py3 as it's not a 1 to 1 backport of selectors from py3.4+.  From the docs:

> All the additional features that selectors2 provides are real-world problems that have occurred and been reported during the lifetime of its maintenance and use within urllib3.

So I'm guessing that one of those additional features, might have resolved the problem.

Can you of you test a change similar to:

```diff
diff --git a/lib/ansible/compat/selectors/__init__.py b/lib/ansible/compat/selectors/__init__.py
index 6458438d6d..42cee28eb8 100644
--- a/lib/ansible/compat/selectors/__init__.py
+++ b/lib/ansible/compat/selectors/__init__.py
@@ -30,14 +30,10 @@ import os.path
 import sys
 
 try:
-    # Python 3.4+
-    import selectors as _system_selectors
+    # backport package installed in the system
+    import selectors2 as _system_selectors
 except ImportError:
-    try:
-        # backport package installed in the system
-        import selectors2 as _system_selectors
-    except ImportError:
-        _system_selectors = None
+    _system_selectors = None
 
 if _system_selectors:
     selectors = _system_selectors
```

needs_info

**Comment by mildebrandt** [comment]
@sivel I should have mentioned, I tried that too with no success. Good idea though! :)

**Comment by ksmets** [comment]
I confirm this issue also exist when running ansible on Ubuntu 16.04 in python 3 environment. Disabling pipelining provides a workaround

**Comment by sivel** [comment]
I've been able to bisect this issue to f921369445900dcc756821e40c9257d5359087af

I've not determined the specific cause within that commit yet.

**Comment by sivel** [comment]
Ok, that commit wasn't it.  It actually addressed a bug that I was "utilizing" in my testing.  I'll go back to it.

**Comment by sivel** [comment]
Ok, good news is that I've actually figured it out.  I'll work on a pull request tomorrow and will link here with that resolution.

In my testing this has existed, at minimum, back to 2.3.0 with python3.

**Comment by sivel** [comment]
I've submitted my pull request, which should resolve these issues.  If anyone could please test it would be appreciated:

https://github.com/ansible/ansible/pull/35049

**Comment by mildebrandt** [comment]
Yes, it worked. Thanks for the quick turnaround!

**Comment by mildebrandt** [comment]
Any chance this can sneak into 2.4.3.0? I see that rc2 was just cut yesterday....so it might be late. Just hoping.

**Comment by sivel** [comment]
It will be included in 2.4.3.  We cherry picked it today, and plan to release a 2.4.3rc3 today.

**Comment by mildebrandt** [comment]
That's great, thanks!

**Comment by abadger** [comment]
Been one of those days....

After some discussion, the bug that was prompting an rc3 release has proven to change behaviour in undesirable ways so it has been reverted. Since there's (at present) no bug that's blocking 2.4.3, there may not be an rc3 and this may not make it into a release until 2.4.4.  I've added the PR to the 2.4.x Project so that if there is a 2.4.3-rc3, I do not forget to re-add it to the branch.
